spring-cloud-stream-binders/spring-cloud-stream-binder-gemfire/src/main/java/org/springframework/cloud/stream/binder/gemfire/GemfireMessageChannelBinder.java

```diff
@@ -0,0 +1,338 @@
+		logger.debug("bindProducer({})", name);
+		Assert.isInstanceOf(SubscribableChannel.class, outboundBindTarget);
+
+		((SubscribableChannel) outboundBindTarget).subscribe(
+				new SendingHandler(createProducerMessageRegion(name),
+						this.cache.getDistributedSystem().getDistributedMember().getProcessId()));
+	}
+
+	@Override
+	public void bindPubSubProducer(String name, MessageChannel outboundBindTarget, Properties properties) {
+	}
+
+	@Override
+	public void bindRequestor(String name, MessageChannel requests, MessageChannel replies, Properties properties) {
+	}
+
+	@Override
+	public void bindReplier(String name, MessageChannel requests, MessageChannel replies, Properties properties) {
+	}
+
+
+	/**
+	 * {@link com.gemstone.gemfire.cache.CacheListener} implementation that
+	 * {@link Object#notifyAll() notifies} itself when a new entry is added
+	 * to the region it is registered for.
+	 */
+	private static class MessageListener extends CacheListenerAdapter<MessageKey, Message<?>> {
+
+		public MessageListener() {
+		}
+
+		@Override
+		public synchronized void afterCreate(EntryEvent<MessageKey, Message<?>> event) {
+			this.notifyAll();
+		}
+	}
+
+
+	/**
+	 * Reads {@link Message} objects from a {@link Region} and
+	 * publishes them to a {@link MessageChannel}.
+	 */
+	private static class QueueReader implements Runnable {
+
+		private final Region<MessageKey, Message<?>> messageRegion;
+
+		private final MessageChannel messageChannel;
+
+		private final MessageListener messageListener;
+
+		private final int batchSize;
+
+		private final boolean inOrder;
+
+		private volatile boolean running = true;
+
+		public QueueReader(Region<MessageKey, Message<?>> messageRegion,
+				MessageChannel messageChannel, MessageListener messageListener,
+				int batchSize, boolean inOrder) {
+			this.messageRegion = messageRegion;
+			this.messageChannel = messageChannel;
+			this.messageListener = messageListener;
+			this.batchSize = batchSize;
+			this.inOrder = inOrder;
+		}
+
+		public boolean isRunning() {
+			return this.running;
+		}
+
+		public void shutdown() {
+			this.running = false;
+		}
+
+		@Override
+		public void run() {
+			// only messages that are present in this JVM will be processed
+			Region<MessageKey, Message<?>> localMessageRegion =
+					PartitionRegionHelper.getLocalData(this.messageRegion);
+
+			while (isRunning()) {
+				if (localMessageRegion.isEmpty()) {
+					try {
+						synchronized (this.messageListener) {
+							this.messageListener.wait(100);
+						}
+					}
+					catch (InterruptedException e) {
+						logger.warn("Thread interrupted", e);
+						Thread.currentThread().interrupt();
+						shutdown();
+					}
+				}
+				else {
+					List<MessageKey> keys = new ArrayList<>(localMessageRegion.keySet());
+					logger.debug("Fetched {} messages", keys.size());
+
+					if (this.inOrder) {
+						Collections.sort(keys);
+					}
```

Comment:
Is there any other kind of ordering other than what a single thread in a producer creates? Otherwise we'd be talking about global ordering, which means a single point of bottleneck for new messages.